src/poetry/core/constraints/version/version_range.py

```diff
@@ -143,11 +150,21 @@ def intersect(self, other: VersionConstraint) -> VersionConstraint:
         if isinstance(other, VersionUnion):
             return other.intersect(self)
 
-        # A range and a Version just yields the version if it's in the range.
         if isinstance(other, Version):
+            # A range and a Version just yields the version if it's in the range.
             if self.allows(other):
                 return other
 
+            # `>=1.2.3+local` intersects `1.2.3` to return `>=1.2.3+local,<1.2.4`.
+            if self.min is not None and self.min.is_local() and other.allows(self.min):
+                upper = other.stable.next_patch()
```

Comment:
```suggestion
                upper = other.stable.next()
```

`next_patch()` will only work for versions with 3 parts, won't it? We should probably add a test with a version with less parts and another test with more parts.

Reply:
`next_patc()` is correct for versions iwth fewer than three parts, no?

Happy to see someone submit a follow-up to deal with 1.2.3.4+local, meanwhile this seems like an edge case that I don't care to worry about...

Reply:
I see it now. Strictly speaking, you can't specify the upper bound. Independent from the precision of the version it's always like `major.minor.patch. ... .0.0.0. ... .1`. In the example, we mistakenly allow `1.2.3.1`. That's probably something we have to accept at a certain precision...

Reply:
Yeah, if y'all want to insist that this MR makes everything perfect then sorry: not gonna happen.  Feel free to make it better or to follow up with MRs that improve it.

My position is that this is a definite improvement, solving an actual problem, and that's what I have the time / energy to do right now.  Take it or leave it!

Reply:
I was just trying to say that a perfect solution might not be possible with our concept of version ranges. I totally agree that this MR is an improvement.